### PR TITLE
Support for changing ACCESS METHOD from AO to heap

### PR DESCRIPTION
**Pull request by huansong**
Reference:
* Heap to AO: https://github.com/greenplum-db/gpdb/pull/13571
* Comments about the general WF of ATAM from/to AO/AOCO tables: https://github.com/greenplum-db/gpdb/pull/13744

-------------

Main considerations of changes:
* Dropping pg_appendonly entries and AO aux tables:
  * We rely on the existing AT framework to drop those since it needs to cleanup the intermediate temporary table in the end. The only thing we need to do is to transfer the AO entry for the original AO table to the temporary heap table and also link the AO aux table dependencies to it too. 
  * So, we can reuse the existing `ATAOEntriesHeapToAO()` function but rename it accordingly.
  * Added more comment for `ATAOEntries` too.
* Reloptions from the old AO are not applicable to new heap table, discard them
* The new heap table should have valid pg_class.relfrozenxid

For `ALTER TABLE SET WITH` from AO to Heap tables, it's trivially supported since the syntax is there. Just adding tests.

### CODE REVIEW

File: src/backend/catalog/pg_appendonly.c [hunk 1]
```diff
@@ -557,9 +560,7 @@ ATAOEntries(Form_pg_class relform1, Form_pg_class relform2)
 			switch(relform2->relam)
 			{
 				case HEAP_TABLE_AM_OID:
-					ereport(ERROR,
-							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-								errmsg("alter table does not support switch from AO to Heap")));
+					TransferAppendonlyEntries(relform1->oid, relform2->oid);
```
Comment (Aegeaner) [inline src/backend/catalog/pg_appendonly.c:563]: Could ACCESS METHOD changed from AOCS to heap too?

**Review by soumyadeep2007 — COMMENTED** [review]
Looking great. I have some comments.

One overall testing suggestion which I think is key (which should be a predecessor commit I feel):
We should `select relname, relam from pg_class where relname in (....);` after each AT AM change operation in our tests for additional validation

**Comment by huansong** [comment]
@soumyadeep2007 Thanks for the much refined comments! All done.

> One overall testing suggestion which I think is key (which should be a predecessor commit I feel):
We should select relname, relam from pg_class where relname in (....); after each AT AM change operation in our tests for additional validation

Done as an additional commit.

File: src/test/regress/expected/alter_table_set_am.out [hunk 2]
```diff
@@ -51,6 +51,14 @@ SET gp_default_storage_options = 'blocksize=65536, compresstype=zlib, compressle
 -- Alter table heap to AO should work
 ALTER TABLE heap2ao SET ACCESS METHOD ao_row;
 ALTER TABLE heap2ao2 SET WITH (appendoptimized=true);
+-- The altered table should have AO AM
```
Comment (soumyadeep2007) [inline src/test/regress/expected/alter_table_set_am.out:54]: Thanks for adding these, we should do the same checks for all other ALTERs in this file, like inherited tables as well.

Comment (soumyadeep2007) [inline src/test/regress/expected/alter_table_set_am.out:54]: s/table/tables

**Review by soumyadeep2007 — APPROVED** [review]
LGTM! One minor outstanding test suggestion.

**Pull request merged**